oximeter/db/src/native/io/column.rs

```diff
@@ -0,0 +1,616 @@
+//
+// Copyright 2024 Oxide Computer Company
+
+//! Encode / decode a column.
+
+use crate::native::{
+    block::{Column, DataType, ValueArray},
+    io, Error,
+};
+use bytes::{Buf as _, BufMut as _, BytesMut};
+use chrono::DateTime;
+use std::net::{Ipv4Addr, Ipv6Addr};
+use uuid::Uuid;
+
+/// Helper macro to quickly and unsafely copy POD data from a message from the
+/// ClickHouse server into our own column data types.
+macro_rules! copyin_pod_values_raw {
+    ($data_type:ty, $src:expr, $n_rows:expr) => {{
+        // Compute the number of total bytes, and try to split the source
+        // buffer at that value. We fail if we have a short read.
+        let n_bytes = $n_rows * core::mem::size_of::<$data_type>();
+        let Some((data, rest)) = $src.split_at_checked(n_bytes) else {
+            return Ok(None);
+        };
+
+        // Preallocate a vec with the data, and create a MaybeUninit wrapper
+        // around it.
+        //
+        // This is used to signal to the compiler that the values here are
+        // uninitialized at first, though they point to valid and aligned
+        // regions of memory.
+        let mut values = Vec::<$data_type>::with_capacity($n_rows);
+        let remaining = values.spare_capacity_mut();
+
+        // Memcpy the data from the I/O buffer into our array.
+        //
+        // Safety: `copy_nonoverlapping` has a lot of invariants that it
+        // expects.
+        //
+        // - Lengths: We have checked the length of the source buffer above, and
+        // we've allocated the destination just above as well.
+        // - Overlap: These cannot overlap, because we've just allocated the new
+        // destination above.
+        // - Alignment: This one is important. We're copying a buf of u8 into
+        // something with possibly larger alignment. So the source buffer may
+        // not have alignment of the destination. However, the reverse is always
+        // true: the destination buffer is always aligned to a u8 buffer,
+        // because all destination types are >= 1 byte. So the ordering of the
+        // calls to `cast()` below is critical, we _must_ cast the destination
+        // values to `*mut u8`, not the source to `*const T`. Note that because
+        // of this, the count argument is in _bytes_, not rows.
+        unsafe {
+            core::ptr::copy_nonoverlapping(
+                data.as_ptr(),
+                remaining.as_mut_ptr().cast(),
+                n_bytes,
+            );
+        }
+
+        // Unsafely set the length to the number of values we copied.
+        //
+        // Safety: This is used for POD types, so any bit pattern should be
+        // valid. But even if that were not true, just above we overwrote every
+        // element, in the call to `copy_nonoverlapping`.
+        unsafe {
+            values.set_len($n_rows);
+        }
+
+        // Chop the prefix out of the source buffer.
+        *$src = rest;
```

Comment:
Great comments above!. I read the documentation for the called methods and I believe everything you did above looks correct. However, I'm not quite sure about this line. I think it's correct, because I don't see how you could write this type of code without doing it this way, but I'm not sure. Since it's a borrow and a pointer, I assume that the owned type will free all the memory since it knows when the original borrow goes out of scope, and that it won't just free some of it, or free it at the wrong time. I'd have someone like @hawkw or @cbiffle take a look, since they have expertise with unsafe.

Reply:
Thanks Andrew! To clarify about this line -- `src` has type `&mut &[u8]`, so it's a mutable reference to an immutable slice of bytes. What we're doing here is effectively moving the byte slice that `src` points to, so that it now points to the first byte after the `memcpy`d data. That's what `rest` is assigned to a few lines above.

Since all these are borrows onto the bytes, I don't think anything is dropped at all when they go out of scope. The references are dropped, but they can't drop anything since they don't own anything.

I would certainly love as many eyes on this as required to prove that, though :)

Reply:
Thanks for pointing that type out. That all makes sense to me.